### PR TITLE
feat(compiler): ECMAScript number formatting parity with fuzz corpus

### LINKED ISSUE #297: compiler M0.3 — Number formatting parity: fuzz String(x) vs Go runtime
## Parent and status — refreshed 2026-09-22

Part of #294. Implementation is in **open PR #304**, head `4f70c0a4e481ea5399972e26fcd8d1bc5c727988`, above #303. Not merged; do not duplicate it. The formatter itself is already in the language-surface layer; this layer supplies deterministic parity coverage and associated scope refinements. #294 records integration status; no tests were rerun for this update.

## What to build

Preserve ECMAScript number-to-string behavior for float64 values in the Go runtime. Reuse Go's shortest-round-trip digit generation; apply ECMAScript decimal/exponent layout and special-value rules rather than writing another dtoa implementation. The current implementation uses `strconv.FormatFloat` with exponential formatting and re-lays out its digits.

For finite nonzero magnitudes, decimal notation applies from `1e-6` inclusive to `1e21` exclusive; exponent notation applies outside that interval. Cover exponent spelling, integer-valued floats, subnormals, shortest-round-trip cases, `NaN` and infinities.

**Distinguish coercion from logging:** JS string coercion of negative zero yields `"0"`; direct `console.log(-0)` yields `-0`. The corpus checks both numeric logging and template coercion. This does not add arbitrary `String()` calls or direct array inspection to the supported language.

## Deterministic parity gate

Use the public `compile()` seam, run the generated program and the Node reference, and compare behavior. The property-style number test is inside the uncached native `test:corpus` lane, not a separate framework or a Go-source snapshot test.

Current pinned corpus:

- Default seed: `0xc2a57`; override with `CRUST_NUMBER_FUZZ_SEED` to replay another run.
- 2,048 generated float64 samples plus 15 curated edge values.
- On an output-parity failure, print the replay seed.
- Generate special values and negative zero with supported runtime expressions so constant folding does not replace the behavior being tested.

These counts describe the current fixture, not a coverage proof over every float64. Old successful runs are historical until rerun on the refreshed stack.

## Acceptance criteria

- [ ] Runtime numeric string coercion follows ECMAScript Number::toString rules for float64; deterministic generated samples and curated edge cases match Node, including thresholds, ±0, subnormals, `NaN` and ±Infinity.
- [ ] Direct numeric logging is also compared, preserving its negative-zero distinction.
- [ ] Seeded generation is reproducible, the environment override works, and parity failures report the seed.
- [ ] The test runs in `test:corpus` and uses public `compile()`; local missing-Go skips are explicit, not passes.
- [ ] There is one runtime formatter path for each required semantic operation, not a retained naive alternative or hand-written dtoa implementation.

## Dependencies

Logical predecessor: #296. PR ancestry: #300 → #303 → #304. New interpreter/runtime profiles are post-M0 work in #294 and do not weaken native number parity.

[Runtime formatting](https://github.com/chenxin-yan/crust/blob/340fc4e4fbf5735cfb4be3e8d7bf9737a2f63aef/packages/compiler/runtime/runtime.go) · [seeded corpus](https://github.com/chenxin-yan/crust/blob/340fc4e4fbf5735cfb4be3e8d7bf9737a2f63aef/packages/compiler/tests/corpus.test.ts)

### PR DESCRIPTION
**Pull request by chenxin-yan**
## Summary

- Verify the ECMAScript number-formatting implementation inherited from #303 using a deterministic Node-vs-native differential corpus.
- Exercise 2,048 raw float64 bit patterns plus 15 curated edge cases, with both direct number logging and template string conversion. Override the seed with `CRUST_NUMBER_FUZZ_SEED`.
- Cover decimal/exponential thresholds, signed exponents, negative zero, NaN, infinities and subnormals without implementing a custom dtoa algorithm.
- Defer direct string-array console inspection: remove the Node util.inspect emulation and reject direct array logging through the public compile() API. Retain array coercion in templates and string concatenation.
- Restore six array-rejection cases; the former array-log fixture is used again. Original review history is retained, but array inspection is no longer in the final implementation.

## Stack

`main` → #300 → #303 → #304 → #302 → #301

Restacked on main `975d893ebf0f7e1e36b7262992a9af3ecde54c18`. This PR's verified head is `4f70c0a4e481ea5399972e26fcd8d1bc5c727988`. All 50 original commits are retained across the stack, plus eight scoped fixes; no merge commits.

## Validation

- Guarded `bun install --frozen-lockfile`, with unchanged lockfile and release-age policy.
- Compiler build, typecheck, lint and format checks pass.
- 41 package tests pass on this exact head.
- Built compiler entrypoint works from an external cwd under Bun and Node 24.
- Go-free API/diagnostic checks and explicit native-suite skip behavior verified.
- Native tests use Go 1.26.5 and Node 24.20.0; number-parity layers pass both default and alternate fuzz seeds.
- Final stack passes forced, zero-cache-hit root-equivalent build/lint/format/types/tests and Bun/Node/Deno distribution smokes. Framework tests: 1,699 pass and seven unrelated conditional skips.
- Independent correctness and standards reviews found no remaining blocking source defects. Hosted CI is separate from this local evidence.

Closes #297

### CODE REVIEW

**Comment by chenxin-yan** [comment]
[stack] Merge order: #300 -> #303 -> #304 -> #302 -> #301 (part 3/5). Stacked PRs get no framework CI until retargeted to main on parent merge; all gates were run locally on the integrated stack tip.

File: packages/compiler/runtime/runtime.go [hunk 1]
```diff
@@ -18,7 +19,17 @@ func Exit(code float64) {
...
 func Log(values ...any) {
-	fmt.Println(values...)
+	for index, value := range values {
+		if index > 0 {
+			fmt.Print(" ")
+		}
+		if number, ok := value.(float64); ok && number == 0 && math.Signbit(number) {
+			fmt.Print("-0")
+		} else {
+			fmt.Print(String(value))
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go:96]: [pi-watch] fixed in 2548f4d

[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go:96]: [pi-watch] fixed in 2548f4d

File: packages/compiler/runtime/runtime.go [hunk 2]
```diff
@@ -32,8 +50,63 @@ func String(value any) string {
...
 }
+
+func inspectStringArray(values []string) string {
+	if len(values) == 0 {
+		return "[]"
+	}
+	quoted := make([]string, len(values))
+	for index, value := range values {
+		value = strings.NewReplacer("\\", "\\\\", "'", "\\'", "\n", "\\n", "\r", "\\r", "\t", "\\t").Replace(value)
+		quoted[index] = "'" + value + "'"
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in b774227f — inspected strings now select single, double, or backtick quotes using Node-compatible rules.

File: packages/compiler/runtime/runtime.go [hunk 3]
```diff
@@ -32,8 +50,63 @@ func String(value any) string {
...
 	}
 }
+
+func inspectStringArray(values []string) string {
+	if len(values) == 0 {
+		return "[]"
+	}
+	quoted := make([]string, len(values))
+	for index, value := range values {
+		value = strings.NewReplacer("\\", "\\\\", "'", "\\'", "\n", "\\n", "\r", "\\r", "\t", "\\t").Replace(value)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in b774227f — array inspection now escapes all C0/C1 control characters using Node-compatible spellings, with differential coverage.

File: packages/compiler/runtime/runtime.go [hunk 4]
```diff
@@ -52,8 +70,63 @@ func String(value any) string {
...
+		quoted[index] = "'" + value + "'"
+	}
+	return "[ " + strings.Join(quoted, ", ") + " ]"
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in b774227f — array inspection now applies Node-compatible 80-column wrapping and grouping; long and multi-item cases are covered.

File: packages/compiler/runtime/runtime.go [hunk 5]
```diff
@@ -107,6 +110,133 @@ func String(value any) string {
 	}
 }
 
+func inspectStringArray(values []string) string {
+	if len(values) == 0 {
+		return "[]"
+	}
+	quoted := make([]string, len(values))
+	for index, value := range values {
+		quoted[index] = quoteInspectString(value)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 9426053b

File: packages/compiler/runtime/runtime.go [hunk 6]
```diff
@@ -52,8 +71,178 @@ func String(value any) string {
...
+	}
+	if len(quoted) > 6 {
+		if grouped := groupInspectStrings(quoted); len(grouped) != len(quoted) {
+			return "[\n  " + strings.Join(grouped, ",\n  ") + "\n]"
+		}
+	}
+	if inspectStringsFitLine(quoted) {
+		return "[ " + strings.Join(quoted, ", ") + " ]"
+	}
+	return "[\n  " + strings.Join(quoted, ",\n  ") + "\n]"
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go:187]: [pi-watch] fixed in 9426053b

File: packages/compiler/runtime/runtime.go [hunk 7]
```diff
@@ -107,6 +110,165 @@ func String(value any) string {
...
+	}
+	return true
+}
+
+func groupInspectStrings(values []string) []string {
+	lengths := make([]int, len(values))
+	totalLength := 0
+	maxLength := 0
+	for index, value := range values {
+		length := len(utf16.Encode([]rune(value)))
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 2fba1b56 — inspection grouping now measures wide Unicode display columns.

File: packages/compiler/runtime/runtime.go [hunk 8]
```diff
@@ -115,6 +119,189 @@ func String(value any) string {
...
+			character == 0x2329 || character == 0x232a ||
+			character >= 0x2e80 && character <= 0xa4cf && character != 0x303f ||
+			character >= 0xac00 && character <= 0xd7a3 ||
+			character >= 0xf900 && character <= 0xfaff ||
+			character >= 0xfe10 && character <= 0xfe19 ||
+			character >= 0xfe30 && character <= 0xfe6f ||
+			character >= 0xff00 && character <= 0xff60 ||
+			character >= 0xffe0 && character <= 0xffe6 ||
+			character >= 0x1f300 && character <= 0x1faff ||
+			character >= 0x20000 && character <= 0x3fffd):
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in b85fa529: regional-indicator runes now count as wide, with an eight-flag differential regression case.

File: packages/compiler/runtime/runtime.go [hunk 9]
```diff
@@ -115,6 +119,190 @@ func String(value any) string {
...
+}
+
+func inspectStringsFitLine(values []string) bool {
+	count := len(values)
+	length := 2*count + 11
+	if length+count > 80 {
+		return false
+	}
+	for _, value := range values {
+		length += inspectStringWidth(value)
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in 5d5d7d15: the single-line fit check now uses UTF-16 length while grouping still uses display width.

File: packages/compiler/runtime/runtime.go [hunk 10]
```diff
@@ -115,6 +119,190 @@ func String(value any) string {
...
+}
+
+func inspectStringWidth(value string) int {
+	width := 0
+	for _, character := range value {
+		switch {
+		case unicode.Is(unicode.Mn, character), unicode.Is(unicode.Me, character), character == '\u200d':
+		case character >= 0x1100 && (character <= 0x115f ||
+			character == 0x2329 || character == 0x232a ||
+			character >= 0x2e80 && character <= 0xa4cf && character != 0x303f ||
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in e5bbe94d — split the broad CJK width range around Node-narrow gaps and added U+3248 differential coverage.

File: packages/compiler/runtime/runtime.go [hunk 11]
```diff
@@ -115,6 +119,192 @@ func String(value any) string {
...
+		lines = append(lines, line.String())
+	}
+	return lines
+}
+
+func inspectStringWidth(value string) int {
+	width := 0
+	for _, character := range value {
+		switch {
+		case unicode.Is(unicode.Mn, character), unicode.Is(unicode.Me, character), character == '\u200d':
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in cb1e4628: Unicode Cf format controls are now zero-width, with U+200B differential coverage.

File: packages/compiler/runtime/runtime.go [hunk 12]
```diff
@@ -115,6 +119,192 @@ func String(value any) string {
...
+		lines = append(lines, line.String())
+	}
+	return lines
+}
+
+func inspectStringWidth(value string) int {
+	width := 0
+	for _, character := range value {
+		switch {
+		case unicode.Is(unicode.Mn, character), unicode.Is(unicode.Me, character), unicode.Is(unicode.Cf, character):
```
[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in f547f674: only Node-zero-width U+200B–U+200F controls are ignored; other format characters count toward layout.

[the remark this answers is not in the document]
Comment (chenxin-yan) [reply packages/compiler/runtime/runtime.go]: [pi-watch] fixed in f547f674: joined emoji components are measured as one display sequence, with rainbow-flag coverage.
